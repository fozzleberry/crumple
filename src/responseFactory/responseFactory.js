"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ResponseFactory {
    static createResponse(status, data, statusCode) {
        const response = {
            status,
            data,
            statusCode
        };
        return response;
    }
    static fail(data, res, statusCode = 422) {
        const response = ResponseFactory.createResponse("fail", data, statusCode);
        res.Send(response);
    }
    static error(data, res, statusCode = 500) {
        const response = ResponseFactory.createResponse("error", data, statusCode);
        res.Send(response);
    }
    static success(data, res, statusCode = 200) {
        const response = ResponseFactory.createResponse("success", data, statusCode);
        res.Send(response);
    }
    static responseByStatusCode(data, res, statusCode) {
        if (!statusCode) {
            return null;
        }
        if (statusCode < 400) {
            ResponseFactory.success(data, res, statusCode);
        }
        else {
            ResponseFactory.error(data, res, statusCode);
        }
    }
}
exports.ResponseFactory = ResponseFactory;
