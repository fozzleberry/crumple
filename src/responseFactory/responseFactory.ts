export type linkedResponse = {
  status: string;
  statusCode: number;
  link: string;
  data?: any;
};

export type okResponse = {
  status: string;
  statusCode: number;
  data: any;
  link?: string;
};

export type errorResponse = {
  status: string;
  statusCode: number;
  message: string;
  data?: any;
};

export type errorMessages = {
  notFound: string;
  internalServer: string;
  validationError: string;
  badRequest: string;
  conflict: string;
};

export type statusCodes = {
  success: number;
  created: number;
  badRequest: number;
  notFound: number;
  conflict: number;
  validationError: number;
  internalServer: number;
};

export type statusMessages = {
  success: string;
  fail: string;
  error: string;
};

export interface IResponseFactory {
  errorMessages: errorMessages;
  statusCodes: statusCodes;
  statusMessages: statusMessages;
  ok: (data?: any, link?: any) => okResponse;
  created: (link: string, data?: any) => linkedResponse;
  response: (
    status: string,
    statusCode: string,
    data: any,
    ...others: any[]
  ) => any;
  invalid: (data?: any) => errorResponse;
  notFound: (data?: any) => errorResponse;
  badRequest: (data?: any) => errorResponse;
  internalServerError: (data?: any) => errorResponse;
  conflict: (data?: any) => errorResponse;
  errorResponse: (
    status: string,
    statusCode: string,
    message: string,
    data?: any,
    ...others: any[]
  ) => any;
  [name: string]: any;
}

export class ResponseFactory {
  protected errorMessages: errorMessages = {
    notFound: "not found",
    internalServer: "internal server error",
    validationError: "validation error",
    badRequest: "bad request",
    conflict: "resource already exists ( duplicate key )"
  };

  protected statusCodes: statusCodes = {
    success: 200,
    created: 201,
    badRequest: 400,
    notFound: 404,
    conflict: 409,
    validationError: 422,
    internalServer: 500
  };

  protected statusMessages: statusMessages = {
    success: "success",
    fail: "fail",
    error: "error"
  };

  ok(data?: any, link?: any): okResponse {
    return {
      status: this.statusMessages.success,
      statusCode: this.statusCodes.success,
      data,
      link
    };
  }

  created(link: string, data?: any): linkedResponse {
    return {
      status: this.statusMessages.success,
      statusCode: this.statusCodes.created,
      link,
      data
    };
  }

  response(
    status = this.statusMessages.success,
    statusCode = this.statusCodes.success,
    data: any,
    ...others: any[]
  ): any {
    return {
      status: this.statusMessages.success,
      statusCode: this.statusCodes.success,
      data,
      ...others
    };
  }

  invalid(data?: any): errorResponse {
    return {
      status: this.statusMessages.fail,
      statusCode: this.statusCodes.validationError,
      message: this.errorMessages.validationError,
      data
    };
  }

  notFound(data?: any): errorResponse {
    return {
      status: this.statusMessages.error,
      statusCode: this.statusCodes.notFound,
      message: this.errorMessages.notFound,
      data
    };
  }

  badRequest(data?: any): errorResponse {
    return {
      status: this.statusMessages.error,
      statusCode: this.statusCodes.badRequest,
      message: this.errorMessages.badRequest,
      data
    };
  }

  internalServerError(data?: any): errorResponse {
    return {
      status: this.statusMessages.error,
      statusCode: this.statusCodes.internalServer,
      message: this.errorMessages.internalServer,
      data
    };
  }

  conflict(data?: any): errorResponse {
    return {
      status: this.statusMessages.error,
      statusCode: this.statusCodes.conflict,
      message: this.errorMessages.conflict,
      data
    };
  }

  errorResponse(
    status = this.statusMessages.error,
    statusCode = this.statusCodes.internalServer,
    message = this.errorMessages.internalServer,
    data?: any,
    ...others: any[]
  ): any {
    return {
      status,
      statusCode,
      message,
      data,
      ...others
    };
  }
}
