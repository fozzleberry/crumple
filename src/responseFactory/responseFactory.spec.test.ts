import { ResponseFactory } from "./responseFactory";

let rf: ResponseFactory;

beforeAll(() => {
  rf = new ResponseFactory();
});

describe("ResponseFactory", () => {
  it("should return an ok response", () => {
    const response = rf.ok({ some: "data" }, "someLink");
    expect(response).toMatchSnapshot();
  });

  it("should return a created response", () => {
    const response = rf.created("someLink", { some: "data" });
    expect(response).toMatchSnapshot();
  });

  it("should return a formatted response", () => {
    const response = rf.response("success", 200, { some: "data" });
    expect(response).toMatchSnapshot();
  });

  it("should return an invalid response", () => {
    const response = rf.invalid({ some: "data" });
    expect(response).toMatchSnapshot();
  });

  it("should return a not found response", () => {
    const response = rf.notFound({ some: "data" });
    expect(response).toMatchSnapshot();
  });

  it("should return a bad request response", () => {
    const response = rf.badRequest({ some: "data" });
    expect(response).toMatchSnapshot();
  });

  it("should return a conflict response", () => {
    const response = rf.conflict({ some: "data" });
    expect(response).toMatchSnapshot();
  });

  it("should return an error response", () => {
    const response = rf.errorResponse("error", 413, "some message", {
      some: "data"
    });
    expect(response).toMatchSnapshot();
  });
});
