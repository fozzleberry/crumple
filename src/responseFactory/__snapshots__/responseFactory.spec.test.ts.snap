// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponseFactory should return a bad request response 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "message": "bad request",
  "status": "error",
  "statusCode": 400,
}
`;

exports[`ResponseFactory should return a conflict response 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "message": "resource already exists ( duplicate key )",
  "status": "error",
  "statusCode": 409,
}
`;

exports[`ResponseFactory should return a created response 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "link": "someLink",
  "status": "success",
  "statusCode": 201,
}
`;

exports[`ResponseFactory should return a formatted response 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "status": "success",
  "statusCode": 200,
}
`;

exports[`ResponseFactory should return a not found response 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "message": "not found",
  "status": "error",
  "statusCode": 404,
}
`;

exports[`ResponseFactory should return an error response 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "message": "some message",
  "status": "error",
  "statusCode": 413,
}
`;

exports[`ResponseFactory should return an invalid response 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "message": "validation error",
  "status": "fail",
  "statusCode": 422,
}
`;

exports[`ResponseFactory should return an ok response 1`] = `
Object {
  "data": Object {
    "some": "data",
  },
  "link": "someLink",
  "status": "success",
  "statusCode": 200,
}
`;
